syntax = "proto3";
package metisfl;

import "metisfl/proto/model.proto";
import "metisfl/proto/service_common.proto";

service LearnerService {
  rpc GetHealthStatus (GetHealthStatusRequest) returns (GetHealthStatusResponse) {}

  rpc GetModel (GetModelRequest) returns (GetModelResponse) {}

  rpc SetInitialWeights (SetInitialWeightsRequest) returns (SetInitialWeightsResponse) {}

  rpc Train (TrainRequest) returns (TrainResponse) {}

  rpc Evaluate (EvaluateRequest) returns (EvaluateResponse) {}

  rpc ShutDown (ShutDownRequest) returns (ShutDownResponse) {}
}

message GetModelRequest {}

message GetModelResponse {
  Model model = 1;
}

message SetInitialWeightsRequest {
  Model model = 1;
}

message SetInitialWeightsResponse {
  Model model = 1;
}

message TrainRequest {
  Model model = 1;
  TrainParams params = 2;
}

message TrainParams {
  uint32 global_iteration = 1;
  uint32 batch_size = 2;
  uint32 epochs = 3;
  repeated string metrics = 4;
}

message TrainResponse {}

message EvaluateRequest {
  Model model = 1;
  EvalParams params = 2;  
}

message EvalParams {
  uint32 batch_size = 1;
  repeated string metrics = 2;
}

message EvaluateResponse {
  map<string, float> metrics_values = 1;
}

